packages/runtime/container-runtime/src/test/batching.spec.ts

```diff
@@ -0,0 +1,229 @@
+
+import { strict as assert } from "assert";
+
+import { AttachState, ICriticalContainerError } from "@fluidframework/container-definitions";
+import { IContainerContext } from "@fluidframework/container-definitions/internal";
+import { FluidErrorTypes } from "@fluidframework/core-interfaces/internal";
+import {
+	MessageType,
+	ISequencedDocumentMessage,
+} from "@fluidframework/driver-definitions/internal";
+import { MockLogger, mixinMonitoringContext } from "@fluidframework/telemetry-utils/internal";
+import {
+	MockAudience,
+	MockDeltaManager,
+	MockQuorumClients,
+	validateAssertionError,
+} from "@fluidframework/test-runtime-utils/internal";
+
+import type { ChannelCollection } from "../channelCollection.js";
+import { ContainerRuntime } from "../containerRuntime.js";
+import { ContainerMessageType } from "../messageTypes.js";
+
+describe("Runtime batching", () => {
+	const mockClientId = "mockClientId";
+
+	const getMockContext = (deltaManager: MockDeltaManager): Partial<IContainerContext> => {
+		const mockContext = {
+			attachState: AttachState.Attached,
+			deltaManager,
+			audience: new MockAudience(),
+			quorum: new MockQuorumClients(),
+			taggedLogger: mixinMonitoringContext(new MockLogger()).logger,
+			clientDetails: { capabilities: { interactive: true } },
+			closeFn: (_error?: ICriticalContainerError): void => {},
+			updateDirtyContainerState: (_dirty: boolean) => {},
+			getLoadedFromVersion: () => undefined,
+			clientId: mockClientId,
+			connected: true,
+		};
+		return mockContext;
+	};
+
+	const mockProvideEntryPoint = async () => ({
+		myProp: "myValue",
+	});
+
+	let containerRuntime: ContainerRuntime;
+	let mockDeltaManager: MockDeltaManager;
+
+	beforeEach(async () => {
+		mockDeltaManager = new MockDeltaManager();
+		containerRuntime = await ContainerRuntime.loadRuntime({
+			context: getMockContext(mockDeltaManager) as IContainerContext,
+			registryEntries: [],
+			existing: false,
+			runtimeOptions: {},
+			provideEntryPoint: mockProvideEntryPoint,
+		});
+	});
+
+	/** Overwrites channelCollection property to make process a no-op */
+	function patchContainerRuntime(cr: ContainerRuntime): ContainerRuntime {
+		const patched = cr as unknown as Omit<ContainerRuntime, "channelCollection"> & {
+			channelCollection: Partial<ChannelCollection>;
+		};
+
+		patched.channelCollection = {
+			process: (
+				message: ISequencedDocumentMessage,
+				local: boolean,
+				localMessageMetadata: unknown,
+			) => {},
+		} satisfies Partial<ChannelCollection>;
+		return patched as unknown as ContainerRuntime;
+	}
+
+	/**
+	 * Returns a batch of messages with the first and last message marked as batch start and end respectively.
+	 */
+	function getBatch(count: number): ISequencedDocumentMessage[] {
+		const messages: ISequencedDocumentMessage[] = [];
+		for (let i = 0; i < count; i++) {
+			messages.push({
+				type: MessageType.Operation,
+				clientId: mockClientId,
+				sequenceNumber: i,
+				minimumSequenceNumber: 1,
```

Comment:
I'm sure it won't matter for unit tests, but you have MSN > sequenceNumber which I think is generally invalid